### PR TITLE
Tidy up the Graph component (part #2)

### PR DESCRIPTION
**Pull request by kinow**
These changes partially address #74 and #381, see also #302 

Initial view of the Graph component when using this branch:

image

After dismissing the warning, and expanding the layout options:

image

Now left with a frozen graph, and dagre engine. While right has dismissed the warning, hierarchical layout, layout options hidden, and it is showing live data.

image

Tests above were done with Firefox 72.0.1 (64 bit) on Ubuntu LTS, using my home laptop.

@hjoliver it kinda works with 60.8. Will take another look tomorrow, but I suspect it's the Lumino widget and how it is handling updates and resizing.

For me, the `/#/graph/families2` view worked fine.

image

But got the same issue reported by @hjoliver when using Lumino.

image

**Requirements check-list**
- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Does not need tests (why? it's simply deleting or moving code).
- [x] No change log entry required (why? e.g. invisible to users).
- [x] No documentation update required.

### CODE REVIEW

**Comment by kinow** [comment]
Notes to self:

- I think multiple Graph components with Lumino is broken due to the ID of the DIV container being used... maybe refs will fix it?
- Need to fix the CSS styles that were "fixed" but broke the layout

**Comment by hjoliver** [comment]
> Need to fix the CSS styles that were "fixed" but broke the layout

What do you mean by "the layout" here - the graph layout, or more general page layout of the UI?

**Comment by hjoliver** [comment]
Technically the "graph layout" is the computed relative positions of all the graph nodes - the shape of the graph, sort of ... so maybe you mean the graph styling (and/or the viewport or whatever the term for that is)

**Comment by kinow** [comment]
Oh, yup. Graph styling, sorry. I am just trying to remember where each style is used, and what CSS styles need changing. Should be done by today/tomorrow.

File: src/components/cylc/graph/Graph.vue [hunk 1]
```diff
@@ -0,0 +1,631 @@
+<template>
+  <div class="c-graph">
+    <SyncLoader
+      :loading='loading'
+      :color='vueSpinner.color'
+      :size='vueSpinner.size'
+      class='spinner'
+    ></SyncLoader>
+    <div class='switch-layout ma-0'>
+      <v-expansion-panels
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: Using an expansion panel, I think we have more space to display the graph. It's hidden by default, user can click on the expand button to visualize the settings now.

File: src/components/cylc/graph/Graph.vue [hunk 2]
```diff
@@ -0,0 +1,631 @@
...
+      tippy: null,
+      /**
+       * Graph data. These values are empty by default, then are populated later via watchers.
+       */
+      graphData: {
+        nodes: [],
+        edges: []
+      },
+      cytoscapeInstance: null
+    }
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: Removed data that was not used, or javascript variables declared outside the data (i.e. not reactive, accessible from outside the component), and grouped them with some simple comments for future developers.

File: src/components/cylc/graph/Graph.vue [hunk 3]
```diff
@@ -0,0 +1,631 @@
...
+    this.updateGraph = debounce(this.updateGraph_, 100)
+    this.workflowUpdated = debounce(this.workflowUpdated_, 100)
+  },
+
+  mounted () {
+    // create a div for cytoscape, this is an approach similar to vue-cytoscape
+    // but the vue-cytoscape component offers more than what we need, so we imitate
+    // part of its behaviour here, and ignore the rest and handle the graph and cytoscape
+    // settings ourselves.
+    const element = document.createElement('div')
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: This is the approach of `vue-cytoscape`. We didn't need the whole library, so removed it from `package.json`, and added a few lines here to set up the `<div>` for Cytoscape.

File: src/components/cylc/graph/Graph.vue [hunk 4]
```diff
@@ -0,0 +1,631 @@
...
+    element.setAttribute('class', 'cytoscape')
+    this.$el.appendChild(element)
+    this.cytoscapeInstance = Object.freeze(cytoscape(({
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: Spent half day on this one.

The `data` elements in a Vue app are simple data. No promises of methods being accessible or not.

Furthermore, adding a deeply nested object, means Vue needs to iterate it to add proxies/observers.

This call was crashing my browser :open_mouth: By using `Object.freeze`, Vue is later able to tell you don't want reactivity, doesn't bother setting up reactivity, creating proxy objects, etc.

File: src/components/cylc/graph/Graph.vue [hunk 5]
```diff
@@ -0,0 +1,631 @@
...
+    this.setupHtmlLabel(this.cytoscapeInstance, states)
+    // set the initial zoom-level and pan (position) of the graph
+    // to set only the zoom level, call: this.cytoscapeInstance.zoom(0.5)
+    this.cytoscapeInstance.viewport({
+      zoom: 0.5,
+      pan: {
+        x: 100,
+        y: 100
+      }
+    })
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: Also spent a long time searching for this one. Reaaaally bothered by having to always zoom out when opening the graph for my two simple `five` and `families2` workflows :smile:

File: src/components/cylc/graph/Graph.vue [hunk 6]
```diff
@@ -0,0 +1,631 @@
...
+  },
+
+  beforeDestroy () {
+    // before the component is destroyed, we need to remember to remove the tooltip, otherwise it is still visible
+    // in the next view/route.
+    if (this.tippy !== null) {
+      this.tippy.hide()
+    }
+    this.cytoscapeInstance.panzoom('destroy')
+    document.removeEventListener('keydown', this.keydownEventHandler)
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: These were not being removed. So we had possibly some detached objects & memory leaking somewhere. (Found about the Panzoom clean up call while looking at their docs to set up default values, by accident).

File: src/components/cylc/graph/Graph.vue [hunk 7]
```diff
@@ -0,0 +1,631 @@
...
+  },
+
+  computed: {
+    // we map all the workflows here, but we have the workflowName prop to filter and know which one is being displayed
+    ...mapState('workflows', ['workflows'])
+  },
+
+  methods: {
+    /**
+     * @param {[
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: GitHub syntax parser bug I think.

File: src/components/cylc/graph/Graph.vue [hunk 8]
```diff
@@ -0,0 +1,631 @@
...
+        actions: {}, // actions to be added
+        undoableDrag: true, // Whether dragging nodes are undoable can be a function as well
+        stackSizeLimit: undefined, // Size limit of undo stack, note that the size of redo stack cannot exceed size of undo stack
+        ready: function () {
+          // callback when undo-redo is ready
+          this.loading = false
+        }
+      }
+      instance.undoRedo(undoRedoOptions)
+    },
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: Not quite sure undo-redo is working... but left it anyway. Lumino also has some key listeners...

File: src/components/cylc/graph/Graph.vue [hunk 9]
```diff
@@ -0,0 +1,631 @@
...
+      instance.panzoom(panzoomDefaults)
+    },
+
+    /**
+     * Sets up the navigator plugin in the instance.
+     * @param {cytoscape} instance - the cytoscape instance
+     */
+    setupNavigator (instance) {
+      instance.navigator({
+        container: this.$refs['cytoscape-navigator-overlay'],
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: This was tricky too.

Before, we were accessing the element by CSS class.

When using it with Lumino, you would have two widgets & components. The `.cytoscape-navigator-overlay` would return an array with two elements. So only one would be used by both tabs, the other would appear blank.

File: src/components/cylc/graph/Graph.vue [hunk 10]
```diff
@@ -0,0 +1,631 @@
...
+              const jobId = `<span class="jobid">${job.batchSysJobId}</span>`
+              const jobSquare = `<div class="box-job-graph ${job.state}-graph"></div>`
+              const details = `<details><summary>${jobId}${jobSquare}</summary>`
+              const submitNum = `<br><strong>submit number:  <span style="color: #555;">${job.submitNum}</span></strong>`
+              const batchSysName = `<br><strong>batch sytem name: <span style="color: #555;">${job.batchSysName}</span></strong>`
+              const batchSysJobId = `<br><strong>batch sytem job id: <span style="color: #555;">${job.batchSysJobId}</span></strong>`
+              const host = `<br><strong>host:  <span style="color: #555;">${job.host}</span></strong>`
+              const startedTime = `<br><strong>started time:  <span style="color: #555;">${job.startedTime}</span></strong>`
+              const submittedTime = `<br><strong>submitted time:  <span style="color: #555;">${job.submittedTime}</span></strong>`
+              const finishedTime = `<br><strong>finished time:  <span style="color: #555;">${job.finishedTime}</span></strong><br>`
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: Several string concatenation operations replaced by backticks.

However, we still have a lot of `innerHTML` and styles and attributes being changed directly. This is quite bad for performance, possible to cause reflows too.

- https://coderwall.com/p/o9ws2g/why-you-should-always-append-dom-elements-using-documentfragments
- https://www.sitepoint.com/10-ways-minimize-reflows-improve-performance/

File: src/components/cylc/graph/Graph.vue [hunk 11]
```diff
@@ -0,0 +1,631 @@
...
+    /**
+     * Sets up the event listeners for the instance.
+     * NB: remember to remove the event listeners to avoid memory leaks.
+     */
+    setupEventListeners () {
+      document.addEventListener('keydown', this.keydownEventHandler)
+    },
+
+    keydownEventHandler (event) {
+      if ((event.metaKey && event.key === 'Z') || (event.ctrlKey && event.key === 'Z')) {
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: (`which` is deprecated - https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/which, `key` is a possible replacement... TL;DR: `which` behaviour varies per browser)

**Comment by kinow** [comment]
I think I got too excited, so didn't squash the commits so that I can revert some if requested. Will add some comments to the new code, but here are the main changes:

- There was some code from `vue-cytoscape` ported to the project.
- Replaced string concatenation in the component by backtick (template literal)
- Moved configuration objects to an `index.js` file
- Moved component and related files to its own folder
- Removed unnecessary files
- Fixed typos and unnecessary or wrong CSS styles
- Added comments to code
- Used a `v-switch` and a `v-select` for layout controls (I think @oliver-sanders suggested it in some issue on github or message on riot :tada: )
- Used a `v-expansion-panel` to display the layout options only when requested (more visible space for the graph)
- Set transparency to the minimap for the same reason above
- Defined initial zoom level, and pan (position of graph)
- Changed styles to use as much viewport as possible (both Graph view and on Lumino as widget)
- Separated DOM elements, so that Lumino widgets are now able to display multiple Graph components
- Added some code to clean up event listeners and Cytoscape objects
- A few more changes here and there where I thought the code could be simpler or the IDE raised some warning

File: src/components/cylc/graph/Graph.vue [hunk 12]
```diff
@@ -0,0 +1,631 @@
...
+  }
+}
+</script>
+
+<style lang="scss">
+@import '~@/styles/cytoscape/html-label.scss';
+@import '~@/styles/cytoscape/cytoscape-custom.scss';
+@import '~@/styles/cytoscape/panzoom.css';
+@import '~tippy.js/themes/light-border.css';
+</style>
```
Comment (kinow) [inline src/components/cylc/graph/Graph.vue]: There were 2 blocks of `css` and `scss`, but `scss` can handle `css` too. The Toppy CSS was imported in the JS part (webpack is able to handle that, but easier to have that here I guess)

File: public/index.html [hunk 13]
```diff
@@ -4,7 +4,6 @@
     <meta charset="utf-8">
     <meta name="viewport" content="width=device-width,initial-scale=1.0">
     <link href='https://fonts.googleapis.com/css?family=Roboto:300,400,500,700,300italic,400italic' rel="stylesheet" type="text/css">
-    <link href="https://use.fontawesome.com/releases/v5.0.13/css/all.css" rel="stylesheet">
```
[the remark this answers is not in the document]
Comment (oliver-sanders) [reply public/index.html]: FYI we import fontawesome in index.html https://github.com/cylc/cylc-ui/blob/master/public/index.html#L7

[the remark this answers is not in the document]
Comment (kinow) [reply public/index.html]: I think you mentioned it somewhere else too. That got me thinking what could be using it, and by coincidence I believe it's only the `Graph` component that was using it for icons.

This code removed in the highlight above is exactly to remove the `public/index.html` import :point_up:

File: src/components/cylc/graph/Graph.vue [hunk 14]
```diff
@@ -0,0 +1,631 @@
...
+                elements.nodes.push({
+                  data: node
+                })
+              })
+            }
+          }
+        }
+        // clear existing data
+        this.graphData.edges.length = 0
+        this.graphData.nodes.length = 0
```
Comment (oliver-sanders) [inline src/components/cylc/graph/Graph.vue]: Wow I didn't know you could modify an array by messing with its length parameter!

File: src/components/cylc/graph/Graph.vue (hunk 8 above)
Comment (oliver-sanders) [reply src/components/cylc/graph/Graph.vue]: What is is meant to do?

File: src/components/cylc/graph/Graph.vue (hunk 10 above)
Comment (oliver-sanders) [reply src/components/cylc/graph/Graph.vue]: It's nasty but this code won't make it to 1.0.0 so we can ignore it for now.

File: src/components/cylc/graph/Graph.vue (hunk 8 above)
Comment (kinow) [reply src/components/cylc/graph/Graph.vue]: After you modify the position of nodes, for example, pressing `CTRL + Z` should undo that, and move your node back to its original position.

Demo: https://rawgit.com/iVis-at-Bilkent/cytoscape.js-undo-redo/master/demo.html

File: src/components/cylc/graph/Graph.vue (hunk 11 above)
Comment (oliver-sanders) [reply src/components/cylc/graph/Graph.vue]: Vue has some support for modifier keys too https://vuejs.org/v2/guide/events.html#System-Modifier-Keys

File: src/components/cylc/graph/Graph.vue (hunk 14 above)
Comment (kinow) [reply src/components/cylc/graph/Graph.vue]: Couldn't find a more elegant way to clear lists in JS, but I guess that's normal for JS, I think :grin:

File: src/components/cylc/graph/Graph.vue (hunk 4 above)
Comment (oliver-sanders) [reply src/components/cylc/graph/Graph.vue]: Could do with sticking a comment about this in the code (we have a bad history with forgetting important information like this).

Comment (kinow) [reply src/components/cylc/graph/Graph.vue]: Good idea. Done!

**Comment by kinow** [comment]
Thanks for the review @oliver-sanders !

**Comment by kinow** [comment]
@oliver-sanders sorry, pushed two more commits :point_up: 

The first is to make the widget wrapper to invoke the `resizeGraph()` function of the `Graph` wrapped component, once the widget has been activated. This fixes the Firefox 60.8 issue reported by @hjoliver in #381 

The second commit is to fix a console error I noticed could happen when the user closed the widget while the layout was "running" (`cytoscape.layout().run()` more or less). There is nothing that the user or us can do about it, so the only option for that case is to ignore the error — though it's still possible to capture the error with the browser console by checking the option to break on any exception.

File: src/components/cylc/graph/Graph.vue (hunk 5 above)
Comment (oliver-sanders) [reply src/components/cylc/graph/Graph.vue]: That's a bit nicer than the (100, 100) hack.

**Comment by oliver-sanders** [comment]
> The second commit is to fix a console error I noticed could happen when the user closed the widget while the layout was "running"

Interesting, I didn't see that.

Code looks good.

**Pull request merged**